packages/h5p-server/src/types.ts

```diff
@@ -543,6 +544,48 @@ export interface IUser {
     type: 'local' | string;
 }
 
+/**
+ * Implementations can implement the IContentUserDataStorage interface and pass it to the constructor of
+ * H5PEditor or H5PPlayer if they wish to keep the user state and allows users to continue where they left off.
+ */
+export interface IContentUserDataStorage {
+    /**
+     * Loads the contentUserData for given contentId, dataType and subContentId - called via GET from https://github.com/h5p/h5p-php-library/blob/master/js/h5p.js#L2416
+     * At the moment it does not seem to work: https://github.com/Lumieducation/H5P-Nodejs-library/issues/1014#issuecomment-968139480
+     * instead the generateContentUserDataIntegration(...) method should be used for integrating the contentUserData object
+     * @param contentId The id of the content to load user data from
+     * @param dataType Used by the h5p.js client
+     * @param subContentId The id provided by the h5p.js client call
+     * @param user The user who owns this object
+     * @returns
+     */
+    loadContentUserData(
+        contentId: ContentId,
+        dataType: string,
+        subContentId: string,
+        user: IUser
+    ): Promise<any>;
+
+    saveContentUserData(
+        contentId: ContentId,
+        dataType: string,
+        subContentId: string,
+        data: any,
+        user: IUser
+    ): Promise<void>;
+
+    /**
+     * Loads the contentUserData for given contentId and builds an array of IContentUserData which is used in the integtration object of the rendered H5P by the H5PPlayer.
```

Comment:
This method should not be part of the Storage interface but in the Manager class, as it applies to all storage classes.

Reply:
The problem is that if it is only in the Manager class you have no idea/indication how many subContentIds there are.

Right now the implementation of `generateContentUserDataIntegration` the `IContentUserDataStorage` has to load all ContentUserData and build an array where the position in the array is equal to the subContentId.

If you use `loadContentUserData` from the Manager you don't know how many subContentIds / dataTypes there are. 

Or am I missing something?

Reply:
In this case the interface should include a method like listByContent(contentId, userId) and the actual integration construction should be done in the Manager.